flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBIncrementalCheckpointUtils.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.runtime.state.KeyGroupRange;
+
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+
+import java.util.List;
+
+/**
+ * Utils for RocksDB Incremental Checkpoint.
+ */
+public class RocksDBIncrementalCheckpointUtils {
+
+	public static void clipDBWithKeyGroupRange(
+		RocksDB db,
+		List<ColumnFamilyHandle> columnFamilyHandles,
+		KeyGroupRange targetGroupRange,
+		KeyGroupRange currentGroupRange,
+		int keyGroupPrefixBytes) throws RocksDBException {
+
+		for (ColumnFamilyHandle columnFamilyHandle : columnFamilyHandles) {
+			if (currentGroupRange.getStartKeyGroup() < targetGroupRange.getStartKeyGroup()) {
+				byte[] beginKey = RocksDBKeySerializationUtils.serializeKeyGroup(
+					currentGroupRange.getStartKeyGroup(), keyGroupPrefixBytes);
+				byte[] endKye = RocksDBKeySerializationUtils.serializeKeyGroup(
+					targetGroupRange.getStartKeyGroup(), keyGroupPrefixBytes);
+				db.deleteRange(columnFamilyHandle, beginKey, endKye);
+			}
+
+			if (currentGroupRange.getEndKeyGroup() > targetGroupRange.getEndKeyGroup()) {
+				byte[] beginKey = RocksDBKeySerializationUtils.serializeKeyGroup(
+					targetGroupRange.getEndKeyGroup() + 1, keyGroupPrefixBytes);
+
+				byte[] endKey = new byte[keyGroupPrefixBytes];
+				for (int i = 0; i < keyGroupPrefixBytes; ++i) {
+					endKey[i] = (byte) (0xFF);
+				}
+				db.deleteRange(columnFamilyHandle, beginKey, endKey);
+			}
+		}
+	}
+
+	public static int evaluateGroupRange(KeyGroupRange range1, KeyGroupRange range2) {
```

Comment:
This method name is not so helpful. It does not tell us anything about what is evaluated. Maybe a javadoc on the public methods would be helpful.

Reply:
I think this code could also be based on something like `KeyGroupRange.getIntersection(KeyGroupRange).getNumberOfKeyGroups()`.

Reply:
Yes, will change it. 👍

Reply:
One more thought: do you think it can make sense to also include the state size of the handle in the evaluation score? Only problem here is, is a higher or a lower size better? A higher size could also just mean that the initial database was just not in a well compacted state.

Reply:
I think I'm a bit torn here.

Reply:
Then let's just keep it simple for now, and we can still improve it if we later find that the size can also be an indicator of the better initial db state.

Reply:
Ok